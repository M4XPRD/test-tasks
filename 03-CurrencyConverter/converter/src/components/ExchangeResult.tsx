import { Box, Typography } from '@mui/material';
import { useDispatch, useSelector } from 'react-redux';
import { useEffect, useState } from 'react';
import axios from 'axios';
import { RootState } from '../slices';
import { setExchangeCurrency } from '../slices/currenciesSlice';

const ExchangeResult = () => {
  const [axiosError, setAxisError] = useState(false);
  const dispatch = useDispatch();
  const currencyAmount = useSelector((state: RootState) => state.currencies.currencyAmount);
  const exchangeAmount = useSelector((state: RootState) => state.currencies.exchangeAmount);
  const exchangeFromShort = useSelector((state: RootState) => state.currencies.exchangeFromShort);
  const exchangeToShort = useSelector((state: RootState) => state.currencies.exchangeToShort);
  const exchangeAPI = useSelector((state: RootState) => state.currencies.exchangeRateURL);
  const APIkey = process.env.REACT_APP_CURRENCY_KEY;

  useEffect(() => {
    const getExchange = async () => {
      try {
        const countExchange = await axios.get(exchangeAPI, {
          params: {
            apikey: APIkey,
            base_currency: exchangeFromShort,
            currencies: exchangeToShort,
          },
        });
        const { data } = countExchange.data;
        const result = data[exchangeToShort] * Number(currencyAmount);
        dispatch(setExchangeCurrency(result.toFixed(2)));
        setAxisError(false);
      } catch (e) {
        setAxisError(true);
      }
    };
    // getExchange();
  }, [currencyAmount, exchangeFromShort, exchangeToShort]);

  return (
    <Box className="exchange-result-container">
      {axiosError ? (
        <>
          <Typography>Произошла ошибка</Typography>
          <Typography variant="h6">Попробуйте другую валюту</Typography>
        </>
      ) : (
        <>
          <Typography>
            {currencyAmount}
            {' '}
            {exchangeFromShort}
            {' '}
            =
          </Typography>
          <Typography variant="h5">
            {exchangeAmount}
            {' '}
            {exchangeToShort}
          </Typography>
        </>
      )}
    </Box>
  );
};

export default ExchangeResult;
