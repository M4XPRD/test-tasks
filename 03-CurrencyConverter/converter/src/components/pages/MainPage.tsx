import { useEffect } from 'react';
import { Container, Grid } from '@mui/material';
import { useSelector, useDispatch } from 'react-redux';
import { ThunkDispatch, AnyAction } from '@reduxjs/toolkit';
import { useNavigate } from 'react-router-dom';
import InputAmount from '../InputAmount';
import SelectCountry from '../SelectCountry';
import SwitchCurrency from '../SwitchCurrency';
import { getCountriesList } from '../../slices/countriesSlice';
import { AppDispatch, RootState } from '../../slices';
import AnimationBar from '../AnimationBar';
import { setExchange } from '../../slices/currenciesSlice';
import ExchangeResult from '../ExchangeResult';
import routes from '../../routes';

const MainPage = () => {
  const dispatch = useDispatch<ThunkDispatch<RootState, AppDispatch, AnyAction>>();
  const navigate = useNavigate();
  const loadingStatus = useSelector((state: RootState) => state.countries.loadingStatus);
  const exchangeFrom = useSelector((state: RootState) => state.currencies.exchangeFrom);
  const exchangeTo = useSelector((state: RootState) => state.currencies.exchangeTo);

  useEffect(() => {
    dispatch(getCountriesList());
  }, []);

  return (
    <div className="main-content">
      <h1 className="converter-label">Конвертер валют</h1>
      {loadingStatus !== 'finished'
        ? <AnimationBar /> : (
          <Container maxWidth="md" className="currency-container">
            <Grid container spacing={2} className="grid-container">
              <InputAmount />
              <SelectCountry currencyValue={exchangeFrom} label="У меня есть" setExchange={setExchange} />
              <SwitchCurrency />
              <SelectCountry currencyValue={exchangeTo} label="Меняю на" setExchange={setExchange} />
            </Grid>
            <ExchangeResult />
          </Container>
        )}
      <button className="to-currencies-button" type="button" onClick={() => navigate(routes.currencies())}>Все курсы валют</button>
    </div>
  );
};

export default MainPage;
