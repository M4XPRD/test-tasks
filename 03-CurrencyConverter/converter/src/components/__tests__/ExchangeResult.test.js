import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../../slices/index';
import ExchangeResult from '../ExchangeResult';
import MainPage from '../pages/MainPage';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}));

describe('ExchangeResult component', () => {
  it('ExchangeResult snapshot', () => {
    const element = render(
      <Provider store={store}>
        <ExchangeResult />
      </Provider>,
    );

    expect(element).toMatchSnapshot();
  });

  it('Should show the currency exchange results', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <MainPage />
        </BrowserRouter>
      </Provider>,
    );
    await waitFor(() => {
      const input = screen.queryByTestId('input-amount').querySelector('input');
      fireEvent.change(input, { target: { value: '1' } });
    });
    await waitFor(() => {
      const dispatchSpy = jest.spyOn(store, 'dispatch');
      const exchangeResult = screen.queryByTestId('exchange-result');
      if (exchangeResult) {
        expect(dispatchSpy).toHaveBeenCalledWith(setExchangeCurrency(expect.any(String)));
        const result = parseFloat(exchangeResult.textContent);
        expect(result).not.toBe(0);
      }
    });
  });
});
