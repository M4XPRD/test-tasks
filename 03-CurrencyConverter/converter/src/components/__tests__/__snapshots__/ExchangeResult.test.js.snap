// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeResult component ExchangeResult snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="exchange-result-container MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          0
           
          USD
           
          =
        </p>
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          data-testid="exchange-result"
        >
          0
           
          RUB
        </h5>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="exchange-result-container MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        0
         
        USD
         
        =
      </p>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        data-testid="exchange-result"
      >
        0
         
        RUB
      </h5>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
